Frequent failure of TestReduceDataForward_IdleWM unit test

**Describe the bug**
TestReduceDataForward_IdleWM unit test is frequently failing

**To Reproduce**
Steps to reproduce the behavior: 
Go to latest `main` branch and run `go test -run TestReduceDataForward_IdleWM`

Output looks like this:

> jvogelman@macos-VF3V14X2QJ reduce % go test -run TestReduceDataForward_IdleWM
{"level":"info","ts":"2023-07-25T09:16:44.586356-07:00","logger":"numaflow","caller":"publish/publisher.go:71","msg":"Creating a new watermark publisher","entityID":"source-reduce-buffer","otStore":"TestReduceDataForward_IdleWM_OT","hbStore":"TestReduceDataForward_IdleWM_PROCESSORS"}
{"level":"info","ts":"2023-07-25T09:16:44.586981-07:00","logger":"numaflow","caller":"fetch/edge_fetcher.go:56","msg":"Creating a new edge watermark fetcher","bucketName":"source-reduce-buffer"}
{"level":"info","ts":"2023-07-25T09:16:44.587044-07:00","logger":"numaflow","caller":"processor/processor_manager.go:130","msg":"Refreshing ActiveProcessors ticker started"}
{"level":"info","ts":"2023-07-25T09:16:44.587083-07:00","logger":"numaflow","caller":"publish/publisher.go:71","msg":"Creating a new watermark publisher","entityID":"reduce-to-vertex","otStore":"reduce-to-vertex_OT","hbStore":"reduce-to-vertex_PROCESSORS"}
{"level":"error","ts":"2023-07-25T09:16:44.587176-07:00","logger":"numaflow","caller":"processor/processor_manager.go:250","msg":"Unable to find the processor","processorEntity":"source-reduce-buffer","stacktrace":"github.com/numaproj/numaflow/pkg/watermark/processor.(*ProcessorManager).startTimeLineWatcher\n\t/Users/jvogelman/GOPATH/src/github.com/numaproj/numaflow/pkg/watermark/processor/processor_manager.go:250"}
{"level":"info","ts":"2023-07-25T09:16:44.587386-07:00","logger":"numaflow","caller":"processor/processor_manager.go:199","msg":"v.AddProcessor successfully added a new fromProcessor","fromProcessor":"source-reduce-buffer"}
{"level":"info","ts":"2023-07-25T09:16:54.585221-07:00","logger":"numaflow","caller":"inmem/kv_watch.go:79","msg":"stopping watching","pipeline":"testPipeline","bucketName":"TestReduceDataForward_IdleWM_OT","watcher":"TestReduceDataForward_IdleWM_OT"}
{"level":"info","ts":"2023-07-25T09:16:54.585212-07:00","logger":"numaflow","caller":"pnf/ordered.go:174","msg":"forward exiting while waiting for ForwardTask completion event","error":"context deadline exceeded"}
{"level":"info","ts":"2023-07-25T09:16:54.585323-07:00","logger":"numaflow","caller":"inmem/kv_watch.go:79","msg":"stopping watching","pipeline":"testPipeline","bucketName":"TestReduceDataForward_IdleWM_PROCESSORS","watcher":"TestReduceDataForward_IdleWM_PROCESSORS"}
{"level":"info","ts":"2023-07-25T09:16:54.585333-07:00","logger":"numaflow","caller":"pnf/ordered.go:240","msg":"Closed partition writer","bufferTo":"reduce-to-vertex"}
{"level":"info","ts":"2023-07-25T09:16:54.585387-07:00","logger":"numaflow","caller":"reduce/data_forward.go:568","msg":"Stopping reduce data forwarder..."}
{"level":"info","ts":"2023-07-25T09:16:54.585498-07:00","logger":"numaflow","caller":"reduce/data_forward.go:573","msg":"Closed buffer reader","bufferFrom":"source-reduce-buffer"}
{"level":"info","ts":"2023-07-25T09:16:54.585512-07:00","logger":"numaflow","caller":"fetch/edge_fetcher.go:218","msg":"Closing edge watermark fetcher","bucketName":"source-reduce-buffer"}
{"level":"info","ts":"2023-07-25T09:16:54.585547-07:00","logger":"numaflow","caller":"publish/publisher.go:281","msg":"Closing watermark publisher","entityID":"reduce-to-vertex","otStore":"reduce-to-vertex_OT","hbStore":"reduce-to-vertex_PROCESSORS"}
{"level":"error","ts":"2023-07-25T09:16:54.585565-07:00","logger":"numaflow","caller":"publish/publisher.go:296","msg":"Failed to delete the key in the heartbeat bucket","entityID":"reduce-to-vertex","otStore":"reduce-to-vertex_OT","hbStore":"reduce-to-vertex_PROCESSORS","bucket":"reduce-to-vertex_PROCESSORS","key":"reduce-to-vertex","error":"key reduce-to-vertex not found","stacktrace":"github.com/numaproj/numaflow/pkg/watermark/publish.(*publish).Close\n\t/Users/jvogelman/GOPATH/src/github.com/numaproj/numaflow/pkg/watermark/publish/publisher.go:296\ngithub.com/numaproj/numaflow/pkg/reduce.(*DataForward).ShutDown\n\t/Users/jvogelman/GOPATH/src/github.com/numaproj/numaflow/pkg/reduce/data_forward.go:586\ngithub.com/numaproj/numaflow/pkg/reduce.(*DataForward).Start\n\t/Users/jvogelman/GOPATH/src/github.com/numaproj/numaflow/pkg/reduce/data_forward.go:133"}
{"level":"error","ts":"2023-07-25T09:16:54.585919-07:00","logger":"numaflow","caller":"reduce/data_forward.go:587","msg":"Failed to close watermark publisher","error":"key reduce-to-vertex not found","stacktrace":"github.com/numaproj/numaflow/pkg/reduce.(*DataForward).ShutDown\n\t/Users/jvogelman/GOPATH/src/github.com/numaproj/numaflow/pkg/reduce/data_forward.go:587\ngithub.com/numaproj/numaflow/pkg/reduce.(*DataForward).Start\n\t/Users/jvogelman/GOPATH/src/github.com/numaproj/numaflow/pkg/reduce/data_forward.go:133"}
--- FAIL: TestReduceDataForward_IdleWM (10.00s)
    data_forward_test.go:395: expected the toBuffer not to be empty context deadline exceeded



**Environment (please complete the following information):**
Mac M1

**Additional context**
Add any other context about the problem here.

---

<!-- Issue Author: Don't delete this message to encourage other users to support your issue! -->

**Message from the maintainers**:

Impacted by this bug? Give it a 👍. We often sort issues this way to know what to prioritize.

--- a/pkg/reduce/data_forward_test.go
+++ b/pkg/reduce/data_forward_test.go
@@ -1214,7 +1214,7 @@ func fetcherAndPublisher(ctx context.Context, fromBuffer *simplebuffer.InMemoryB
 	// publisher for source
 	sourcePublisher := publish.NewPublish(ctx, sourcePublishEntity, wmstore.BuildWatermarkStore(hb, ot), 1, publish.WithAutoRefreshHeartbeatDisabled())
 
-	// publish heartbeat manually for the processors
+	// publish heartbeat manually for the processor
 	go func() {
 		for {
 			select {
@@ -1231,6 +1231,10 @@ func fetcherAndPublisher(ctx context.Context, fromBuffer *simplebuffer.InMemoryB
 	otWatcher, _ := inmem.NewInMemWatch(ctx, pipelineName, keyspace+"_OT", otWatcherCh)
 	storeWatcher := wmstore.BuildWatermarkStoreWatcher(hbWatcher, otWatcher)
 	pm := processor.NewProcessorManager(ctx, storeWatcher, 1, processor.WithIsReduce(true))
+	for waitForReadyP := pm.GetProcessor(fromBuffer.GetName()); waitForReadyP == nil; waitForReadyP = pm.GetProcessor(fromBuffer.GetName()) {
+		// wait until the test processor has been added to the processor list
+		time.Sleep(time.Millisecond * 100)
+	}
 	f := fetch.NewEdgeFetcher(ctx, fromBuffer.GetName(), storeWatcher, pm, 1)
 	return f, sourcePublisher
 }

--- a/pkg/reduce/data_forward_test.go
+++ b/pkg/reduce/data_forward_test.go
@@ -46,9 +46,11 @@ import (
 	"github.com/numaproj/numaflow/pkg/window/strategy/fixed"
 )
 
+const pipelineName = "testPipeline"
+
 var keyedVertex = &dfv1.VertexInstance{
 	Vertex: &dfv1.Vertex{Spec: dfv1.VertexSpec{
-		PipelineName: "testPipeline",
+		PipelineName: pipelineName,
 		AbstractVertex: dfv1.AbstractVertex{
 			Name: "testVertex",
 			UDF:  &dfv1.UDF{GroupBy: &dfv1.GroupBy{Keyed: true}},
@@ -60,7 +62,7 @@ var keyedVertex = &dfv1.VertexInstance{
 
 var nonKeyedVertex = &dfv1.VertexInstance{
 	Vertex: &dfv1.Vertex{Spec: dfv1.VertexSpec{
-		PipelineName: "testPipeline",
+		PipelineName: pipelineName,
 		AbstractVertex: dfv1.AbstractVertex{
 			Name: "testVertex",
 			UDF:  &dfv1.UDF{GroupBy: &dfv1.GroupBy{Keyed: false}},
@@ -275,7 +277,7 @@ func TestDataForward_StartWithNoOpWM(t *testing.T) {
 	var pbqManager *pbq.Manager
 
 	// create pbqManager
-	pbqManager, err = pbq.NewManager(child, "reduce", "test-pipeline", 0, memory.NewMemoryStores(memory.WithStoreSize(100)),
+	pbqManager, err = pbq.NewManager(child, "reduce", pipelineName, 0, memory.NewMemoryStores(memory.WithStoreSize(100)),
 		pbq.WithReadTimeout(1*time.Second), pbq.WithChannelBufferSize(10))
 	assert.NoError(t, err)
 
@@ -341,7 +343,6 @@ func TestReduceDataForward_IdleWM(t *testing.T) {
 		startTime      = 1679961600000 // time in millis
 		fromBufferName = "source-reduce-buffer"
 		toVertexName   = "reduce-to-vertex"
-		pipelineName   = "test-reduce-pipeline"
 		err            error
 	)
 	defer cancel()
@@ -546,7 +547,6 @@ func TestReduceDataForward_Count(t *testing.T) {
 		startTime      = 60000 // time in millis
 		fromBufferName = "source-reduce-buffer"
 		toVertexName   = "reduce-to-vertex"
-		pipelineName   = "test-reduce-pipeline"
 		err            error
 	)
 
@@ -621,7 +621,6 @@ func TestReduceDataForward_AllowedLatencyCount(t *testing.T) {
 		startTime      = 60000 // time in millis
 		fromBufferName = "source-reduce-buffer"
 		toVertexName   = "reduce-to-vertex"
-		pipelineName   = "test-reduce-pipeline"
 		err            error
 	)
 
@@ -700,7 +699,6 @@ func TestReduceDataForward_Sum(t *testing.T) {
 		startTime      = 0 // time in millis
 		fromBufferName = "source-reduce-buffer"
 		toVertexName   = "reduce-to-vertex"
-		pipelineName   = "test-reduce-pipeline"
 		err            error
 	)
 
@@ -775,7 +773,6 @@ func TestReduceDataForward_Max(t *testing.T) {
 		startTime      = 0 // time in millis
 		fromBufferName = "source-reduce-buffer"
 		toVertexName   = "reduce-to-vertex"
-		pipelineName   = "test-reduce-pipeline"
 		err            error
 	)
 
@@ -851,7 +848,6 @@ func TestReduceDataForward_SumWithDifferentKeys(t *testing.T) {
 		startTime      = 0 // time in millis
 		fromBufferName = "source-reduce-buffer"
 		toVertexName   = "reduce-to-vertex"
-		pipelineName   = "test-reduce-pipeline"
 		err            error
 	)
 
@@ -868,7 +864,7 @@ func TestReduceDataForward_SumWithDifferentKeys(t *testing.T) {
 
 	// create pbq manager
 	var pbqManager *pbq.Manager
-	pbqManager, err = pbq.NewManager(ctx, "reduce", "test-pipeline", 0, memory.NewMemoryStores(memory.WithStoreSize(1000)),
+	pbqManager, err = pbq.NewManager(ctx, "reduce", pipelineName, 0, memory.NewMemoryStores(memory.WithStoreSize(1000)),
 		pbq.WithReadTimeout(1*time.Second), pbq.WithChannelBufferSize(10))
 	assert.NoError(t, err)
 
@@ -947,7 +943,6 @@ func TestReduceDataForward_NonKeyed(t *testing.T) {
 		startTime      = 0 // time in millis
 		fromBufferName = "source-reduce-buffer"
 		toVertexName   = "reduce-to-vertex"
-		pipelineName   = "test-reduce-pipeline"
 		err            error
 	)
 
@@ -1025,7 +1020,6 @@ func TestDataForward_WithContextClose(t *testing.T) {
 		startTime      = 0 // time in millis
 		fromBufferName = "source-reduce-buffer"
 		toVertexName   = "reduce-to-vertex"
-		pipelineName   = "test-reduce-pipeline"
 		err            error
 	)
 
@@ -1048,7 +1042,7 @@ func TestDataForward_WithContextClose(t *testing.T) {
 
 	// create pbq manager
 	var pbqManager *pbq.Manager
-	pbqManager, err = pbq.NewManager(cctx, "reduce", "test-pipeline", 0, storeProvider,
+	pbqManager, err = pbq.NewManager(cctx, "reduce", pipelineName, 0, storeProvider,
 		pbq.WithReadTimeout(1*time.Second), pbq.WithChannelBufferSize(10))
 	assert.NoError(t, err)
 
@@ -1118,7 +1112,6 @@ func TestReduceDataForward_SumMultiPartitions(t *testing.T) {
 		startTime      = 0 // time in millis
 		fromBufferName = "source-reduce-buffer"
 		toVertexName   = "reduce-to-vertex"
-		pipelineName   = "test-reduce-pipeline"
 		err            error
 	)
 
@@ -1136,7 +1129,7 @@ func TestReduceDataForward_SumMultiPartitions(t *testing.T) {
 
 	// create pbq manager
 	var pbqManager *pbq.Manager
-	pbqManager, err = pbq.NewManager(ctx, "reduce", "test-pipeline", 0, memory.NewMemoryStores(memory.WithStoreSize(1000)),
+	pbqManager, err = pbq.NewManager(ctx, "reduce", pipelineName, 0, memory.NewMemoryStores(memory.WithStoreSize(1000)),
 		pbq.WithReadTimeout(1*time.Second), pbq.WithChannelBufferSize(10))
 	assert.NoError(t, err)
 
@@ -1210,7 +1203,6 @@ func fetcherAndPublisher(ctx context.Context, fromBuffer *simplebuffer.InMemoryB
 
 	var (
 		keyspace     = key
-		pipelineName = "testPipeline"
 		hbBucketName = keyspace + "_PROCESSORS"
 		otBucketName = keyspace + "_OT"
 	)
@@ -1220,9 +1212,9 @@ func fetcherAndPublisher(ctx context.Context, fromBuffer *simplebuffer.InMemoryB
 	ot, otWatcherCh, _ := inmem.NewKVInMemKVStore(ctx, pipelineName, otBucketName)
 
 	// publisher for source
-	sourcePublisher := publish.NewPublish(ctx, sourcePublishEntity, wmstore.BuildWatermarkStore(hb, ot), 1, publish.WithAutoRefreshHeartbeatDisabled(), publish.WithPodHeartbeatRate(1))
+	sourcePublisher := publish.NewPublish(ctx, sourcePublishEntity, wmstore.BuildWatermarkStore(hb, ot), 1, publish.WithAutoRefreshHeartbeatDisabled())
 
-	// publish heartbeat for the processors
+	// publish heartbeat manually for the processors
 	go func() {
 		for {
 			select {

